pacopy/natural.py

```diff
@@ -109,5 +118,10 @@ def natural(
 
         callback(k, lmbda, u)
         k += 1
+        if milestones is not None and lmbda == milestone:
```

Comment:
The float comparison is a bit iffy.

Reply:
Yes, interesting.  Usually one avoids testing two floats for equality; here however I think it's O. K.

There are two dangers:
* false negative, missing the milestone; and
* false positive, matching something that isn't the milestone.

I don't think the false positive matters.  If `lmbda` is close enough to the `milestone` to pass an equality test, it's good enough to be considered as it.  I think this is fairly unlikely anyway; I have not been able to conceive a working example.

The false negative one is trickier.  I think it's O. K., this won't happen because it's not really two floats, but, in the case that the milestone is reached, two (shallow) copies of the same float, being one of the specified milestones.  The left term `lmbda` is assigned here (in the case that the milestone is reached) from the second argument of `min`.

https://github.com/nschloe/pacopy/blob/e2d972fcb80c3c43efdeb30e89f78bd83cfd1ad9/pacopy/natural.py#L88

So the question is whether `min` changes its second argument before assigning it to the left-hand side.  It doesn't appear to; i.e.
```python
milestone = 8e2
lmbda = min(1e3, milestone)
assert lmbda == milestone
```
In fact, one even has [object identity](https://docs.python.org/3/reference/expressions.html#is)
```python
assert lmbda is milestone
```
while
```python
assert lmbda is not 8e2
```
Perhaps using `is` here would be clearer?

I don't know.  What do you think?  Is it safe enough?  (Could the behaviour of `min` differ?)  Add a reassuring comment?  

Or is there another way to implement this that avoids the issue?  

I don't think the usual expedient of a tolerance as in [`numpy.allclose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html#numpy.allclose) is appropriate as I can see that producing false negatives which would be messy and annoying. Especially as the other thing is that in the envisaged use-case, the `callback` will always have another such test; e.g.

https://github.com/nschloe/pacopy/blob/e2d972fcb80c3c43efdeb30e89f78bd83cfd1ad9/test/test_bratu1d.py#L106-L107

Reply:
Thinking about this some more.  (Thank you for raising it.  I intend to make much use of this function and feature, so don't mind spending time now getting it right.   I appreciate the rigour.)

Possible 'worst case scenario' (?).  Say in

https://github.com/nschloe/pacopy/blob/98d4ccfa1867d6007c9b2262ccafc18f5fbd5573/pacopy/natural.py#L85-L87

that `lmbda` is numerically close to `milestone` but is judged by `min` to be the lesser.  Then, say we successfully solve at this _&lambda;_ and get to

https://github.com/nschloe/pacopy/blob/98d4ccfa1867d6007c9b2262ccafc18f5fbd5573/pacopy/natural.py#L120

but then fail the second test, the floating-point equality.  What will happen then is that we'll go through the next iteration of the `while`-loop with

https://github.com/nschloe/pacopy/blob/98d4ccfa1867d6007c9b2262ccafc18f5fbd5573/pacopy/natural.py#L85-L87

and this time `lmbda` will go well past `milestone` and so the latter will win the `min` and we'll be trying to solve for `lmbda = milestone` that was very close to the last one.  This will be solved easily by `pacopy.newton` because the previous solution at the close `lmbda` will be an excellent initial guess and we'll be on our way again.   

Side-effect:  `lambda_stepsize` might be increased because `newton_steps` was small.  Hmm.  Is it worth catching or correcting for this?  I don't think that we'd want to reset `lambda_stepsize` when we truncate a step to land on a milestone as the old `lambda_stepsize` is a better estimate of the right one thereafter.

Reply:
The adjustment of `lambda_stepsize` is moved into the `else` clause in ec1b7fc ; i.e. it is not adjusted when the actual step was shortened to meet a milestone.  That solves that last side-effect.